estimator/ntru_primal.py

```
@@ -0,0 +1,503 @@
+from .ntru_parameters import NTRUParameters
+from .simulator import normalize as simulator_normalize
+from .prob import conditional_chi_squared, chisquared_table
+from .io import Logging
+from .conf import red_cost_model as red_cost_model_default
+from .conf import red_shape_model as red_shape_model_default
+from .conf import red_simulator as red_simulator_default
+from .conf import max_n_cache
+
+
+class PrimalDSD():
+    """
+    Estimate cost of solving (overstretched) NTRU via dense sublattice discovery
+    """
+    @staticmethod
+    @cached_function
+    def ball_log_vol(n):
+        return RR((n/2.)) * RR(log(pi)) - RR(lgamma(n/2. + 1))
+
+    @staticmethod
+    def log_gh(d, logvol=0):
+        if d < 49:
+            return RR(gh_constant[d]) + RR(logvol)/d
+
+        return RR(1./d) * RR(logvol - PrimalDSD.ball_log_vol(d))
+
+    @staticmethod
+    def DSL_logvol_matrix(n, sigmasq):
+        total = n*(RR(log(sigmasq))+RR(log(2.))+RR(digamma(n)))/2.
+        proj_loss = np.sum([(digamma((2*n-i)/2.)-digamma(n)) for i in range(n)])/2.
+        return total+proj_loss
+
+    @staticmethod
+    def DSL_logvol_circulant(n, sigmasq):
+        lambda0 = RR((np.log(2)-np.euler_gamma+np.log(n)+np.log(sigmasq))/2.)
+        lambdai = (n-1)*(1-np.euler_gamma+np.log(n)+np.log(sigmasq))/2.
+        return lambda0+lambdai
+
+    @staticmethod
+    def DSL_logvol_circulant_fixed(n, R):
+        lambda0 = (-np.euler_gamma+np.log(R))/2.
+        lambdai = (n-1)*(1-np.euler_gamma+np.log(R)-np.log(2))/2.
+        return lambda0+lambdai
+
+    @staticmethod
+    @cached_function
+    def DSL_logvol(n, sigmasq, ntru="circulant"):
+        if ntru=="matrix":
+            return PrimalDSD.DSL_logvol_matrix(n, sigmasq)
+        if ntru=="circulant":
+            return PrimalDSD.DSL_logvol_circulant(n, sigmasq)
+        if ntru=="fixed":
+            return PrimalDSD.DSL_logvol_circulant_fixed(n, sigmasq)
+        print("non implemented ntru type")
```

You should raise an error here.

---

Fixed in latest commit.